packages/frontend/src/hooks/useKeycard.ts

```diff
@@ -1,27 +1,80 @@
+import { assert } from "@std/assert";
 import { useLocalStorage } from "@uidotdev/usehooks";
 import { generatePrivateKey, privateKeyToAccount } from "viem/accounts";
+import { getLogger } from "@logtape/logtape";
+import { useWalletClient } from "wagmi";
+import { useQuery } from "@tanstack/react-query";
+
+import { RelayClient } from "@massmarket/client";
+import { getWindowLocation } from "@massmarket/utils";
 
 import { useShopId } from "./useShopId.ts";
 import { KeycardRole } from "../types.ts";
+import { useRelayEndpoint } from "./useRelayEndpoint.ts";
+
+type Keycard = {
+  privateKey: `0x${string}`;
+  role: KeycardRole;
+  address: `0x${string}`;
+} | null;
 
-export function useKeycard(
-  keycard?: {
-    privateKey: `0x${string}`;
-    role: KeycardRole;
-    address: `0x${string}`;
-  },
-) {
-  if (!keycard) {
-    const privateKey = generatePrivateKey();
-    const account = privateKeyToAccount(privateKey);
-    keycard = {
-      privateKey,
-      role: KeycardRole.NEW_GUEST,
-      address: account.address,
-    };
-  }
+const logger = getLogger(["mass-market", "frontend", "useKeycard"]);
+
+export function useKeycard(role: KeycardRole = "guest") {
   const { shopId } = useShopId();
-  const keyCardID = "keycard" + shopId;
-  // console.log({shopId, keyCardID});
-  return useLocalStorage(keyCardID, keycard);
+  const { data: wallet } = useWalletClient();
+  const { relayEndpoint } = useRelayEndpoint();
+  const [keycard, setKeycard] = useLocalStorage<Keycard>(
+    "keycard" + shopId,
+    null,
+  );
+
+  return useQuery({
+    queryKey: ["keycard", shopId?.toString(), relayEndpoint, wallet, role],
+    queryFn: async () => {
+      // 1. Generate KC
+      // 2. Enroll KC
+      // 3. Save to localStorage
+
+      // This is for instances when the user has a guest keycard and wants to log in as a merchant.
+      const guestToMerchant = role === "merchant" && keycard?.role === "guest";
+      if (keycard && !guestToMerchant) {
+        return keycard;
```

Comment:
im not sure this works. A merchant can login as a guest though

Reply:
maybe you can use https://tanstack.com/query/latest/docs/reference/QueryClient/#queryclientsetquerydata to set the guest keycard when the merchant is enrolled.